instrumentation/netty-codec-http/src/main/java/brave/netty/http/TracingHttpServerHandler.java

```diff
@@ -0,0 +1,61 @@
+package brave.netty.http;
+
+import brave.Span;
+import brave.Tracer;
+import brave.Tracer.SpanInScope;
+import brave.http.HttpServerHandler;
+import brave.http.HttpTracing;
+import brave.propagation.TraceContext;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+
+final class TracingHttpServerHandler extends ChannelDuplexHandler {
+  final Tracer tracer;
+  final HttpServerHandler<HttpRequest, HttpResponse> handler;
+  final TraceContext.Extractor<HttpHeaders> extractor;
+
+  TracingHttpServerHandler(HttpTracing httpTracing) {
+    tracer = httpTracing.tracing().tracer();
+    handler = HttpServerHandler.create(httpTracing, new HttpNettyAdapter());
+    extractor = httpTracing.tracing().propagation().extractor(HttpHeaders::get);
+  }
+
+  @Override public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+    if (!(msg instanceof HttpRequest)) return;
+
+    HttpRequest httpRequest = (HttpRequest) msg;
+    Span span = handler.handleReceive(extractor, httpRequest.headers(), httpRequest);
+    ctx.channel().attr(NettyHttpTracing.SPAN_ATTRIBUTE).set(span);
+    // TODO figure out how to get the client IP from ctx and set it if not span.isNoop
+
+    try (SpanInScope ws = tracer.withSpanInScope(span)) {
+      super.channelRead(ctx, msg);
```

Comment:
This will never throw

Reply:
Heh umm i suspect someone can write a buggy handle function right?

Reply:
nm sorry mind wasn't operating